### PR TITLE
gh-163: drop `Python 3.9` & `NumPy < 2.1.1` support

### LINKED ISSUE #163: NumPy 2 compatibility (+ drop `3.9` support)
Ensure compatibility with the NumPy 2 release.

### PR DESCRIPTION
**Pull request by paddyroddy**
Closes #163. This will be required for #67. Blocked by #390.

As per the [NumPy 2.1.1 release notes](https://github.com/numpy/numpy/releases/tag/v2.1.1) Python prior to `3.9` is not supported. This release is also the first version with full Array API support. So in order to support it full we need these changes.

There are some code changes due to `ruff` and the minimum Python version being `3.10`.

### CODE REVIEW

File: glass/observations.py [hunk 1]
```diff
@@ -258,7 +251,7 @@ def equal_dens_zbins(
     cuml_nz /= cuml_nz[[-1]]
     zbinedges = np.interp(np.linspace(0, 1, nbins + 1), cuml_nz, z)
 
-    return list(zip(zbinedges, zbinedges[1:]))
+    return list(itertools.pairwise(zbinedges))
```
Comment (Saransh-cpp) [inline glass/observations.py:254]: TIL

**Review by Saransh-cpp — COMMENTED** [review]
Awesome! Will make the Array API transition much smoother

File: glass/observations.py (hunk 1 above)
Comment (paddyroddy) [reply glass/observations.py:254]: You can thank `ruff` https://docs.astral.sh/ruff/rules/zip-instead-of-pairwise

**Review by Saransh-cpp — APPROVED** [review]

File: noxfile.py [hunk 2]
```diff
@@ -10,7 +10,12 @@
 nox.options.reuse_existing_virtualenvs = True
 nox.options.sessions = ["lint", "tests"]
 
-ALL_PYTHON = ["3.9", "3.10", "3.11", "3.12", "3.13"]
+ALL_PYTHON = [
+  "3.10",
+  "3.11",
+  "3.12",
+  "3.13",
+]
```
Comment (paddyroddy) [inline noxfile.py:18]: In future we'll just get one line change rather than many

**Comment by paddyroddy** [comment]
pre-commit.ci autoupdate

**Pull request merged**